### ISSUE TITLE
Prompt user for kernel to use if notebook's metadata is ambiguous (say it's just "Python3")

### ISSUE THREAD

**Issue by greazer** [microsoft/vscode-jupyter#5445]
If we cannot be sure that a kernel installed on a system matches the kernel specified in an open notebook, then we should consider prompting the user to disambiguate. We can make the one we *would* have picked the default. So at worst, it's one extra keystroke when trying to start running a notebook.

**Comment by rchiodo** [comment]
This could be hitting a lot people.

Idea for solution:
- Kernelspec is listed the same as it is now (we don't use the active interpreter if we're doing that now)
- However, when launching the kernel we ask the user to disambiguate at that point
- Then when they make a selection, we write that path to the kernel.json.

**Comment by DonJayamanne** [comment]
What does this fix?
Rather, what's the problem we're trying to fix here?

The assumption here seems to be that we're not starting the right kernel, but what's that based on?

**Comment by rchiodo** [comment]
If the argv of the kernelspec says 'python' which python is it? It might not even be a valid python.

This was the exact problem that Chris Dias hit.

**Comment by DonJayamanne** [comment]
I think we're going to end up with a prompt for a majority of the users who won't need this.
Chris was an execption, and yes we should handle that. 

I believe the root cause of the problem was Chris had Python 2 installed and `python` ended up pointing to `python2`.

Either way, I think the solution is fairly simple:
* Don't display `python` kernels. What's the point of displaying these when using RAW kernel. The kernel spec means absolutely nothing in the kernel spec list, it just points to the interpreter that started jupyter.
    * If the kernel spec doesn't have any env variables, and argv[0] === 'python', then just hide it.
    * This way we force users to select an interpreter or we select the current interpreter by default (which is what we do today).
* If user is using non-Raw, then display these kernels, let Jupyter handle it.
	* & We can easily determine what `python`  points it, it will point to the same interpreter that started Jupyter, we know this.

Basically my goal is try to solve this our selves instead of displaying prompts to the user.
E.g. we can solve the issue that Chris ran into without prompts.

**Comment by rchiodo** [comment]
This situation only occurs when the root jupyter kernelspec (a location where we can't use the relative path to compute the interpreter) points to an invalid or unknown python. The situation Chris had. 

I don't know how often this occurs, but the default jupyter kernelspec would end up being that one. So any normal notebook would reference this kernelspec by name at least.

**Comment by DonJayamanne** [comment]
> This situation only occurs when the root jupyter kernelspe

Then hiding this just makes this go away. Displaying this kernel spec makes absolutely no sense, as the argv == `python` and that could be anything.

I propose we hide this for all new users, and for existing users we add telemetry & slowly hide this for then as well.
Lets discuss in triage

**Comment by rchiodo** [comment]
Notebook says 'Python 3' as the kernelspec. There's only one 'Python 3' kernelspec and it's invalid. What would we show as the kernel to start with?

**Comment by DonJayamanne** [comment]
I'd expect it to point to active interpreter, because that's what we do today, we use the active interpreter as the default kernel.

Hence `python3` or `python` kernels should just point to the active interpreter.
Once the user saves the notebook, this will change to the interpreter path (we update the metadata)

**Comment by rchiodo** [comment]
> I'd expect it to point to active interpreter, because that's what we do today, we use the active interpreter as the default kernel.

Yeah that was the reason for this bug. We felt like this may not be working. Picking the active interpreter may be causing the user problems.

**Comment by DonJayamanne** [comment]
>  We felt like this may not be working. Picking the active interpreter may be causing the user problems.

Chris had an issue because he was having issues with non-existent interpreters, that different from picking active interpreter vs something else.

> . Picking the active interpreter may be causing the user problems.


This is something different. The root cause of the issue is that we were using a non-existen kernel (python2), not sure how we came to the conclusion that using active interrpeter may not mne correct, they are completely diffrerent issues.

**Comment by rchiodo** [comment]
Chris's problem was in addition to the problem of just picking the active interpreter. It was seen as over eager to just pick the active interpreter. Especially for somebody just using notebooks. Is it likely that the active interpreter is the one that created the kernelspec? Not sure we can guess that, hence asking the user. 

I think the first step it to just gather telemetry on how often this occurs. I think we can change it to just using the active interpreter for now (as that at least fixes the problem when 'python' doesn't resolve).

**Comment by DonJayamanne** [comment]
> I think the first step it to just gather telemetry on how often this occurs

I agree, I prefer that over displaying any popups (my fear is that majority of the users will get it when its not requried - again no data for that either).

**Comment by rchiodo** [comment]
This bug sounds similar? At least it doesn't seem like we picked the correct interpreter when we first wrote out the interpreter information in the kernelspec:

https://github.com/microsoft/vscode-jupyter/issues/5518#issuecomment-821279968

**Comment by DonJayamanne** [comment]
> least it doesn't seem like we picked the correct interpreter when we first wrote out the interpreter information in the kernelspec:

I think its different. Its not due to /usr/bin/python, but because we've just picked the wrong interpreter completely. i.e. not caused by some ambiguous interpreter issue there.

**Comment by IanMatthewHuff** [comment]
Validated on default kernel specs not being shown anymore.
